### PR TITLE
WIP Parallel initialization, issue 142, work package 3

### PR DESCRIPTION
**Pull request by atotoun**

### CODE REVIEW

File: src/m2n/M2N.cpp [hunk 1]
```diff
@@ -178,6 +204,29 @@ void M2N::send(double itemToSend)
...
 
+void M2N::broadcastSendLocalMesh(mesh::Mesh &mesh)
+{
+  int meshID = mesh.getID();
+  if (utils::MasterSlave::_slaveMode || utils::MasterSlave::_masterMode) {
+    assertion(_areSlavesConnected);
+    assertion(_distComs.find(meshID) != _distComs.end());
+    assertion(_distComs[meshID].get() != nullptr);    
+    _distComs[meshID]->broadcastSendMesh(mesh);
+  } else { //coupling mode
```
Comment (uekerman) [inline src/m2n/M2N.cpp]: Please add assertion and understandable message

File: src/m2n/PointToPointCommunication.cpp [hunk 2]
```diff
@@ -773,6 +773,34 @@ void PointToPointCommunication::broadcastReceive(double &itemToReceive)
   }  
 }
 
+void PointToPointCommunication::broadcastSendMesh(mesh::Mesh &mesh)
```
Comment (uekerman) [inline src/m2n/PointToPointCommunication.cpp]: The mesh is already a member of this class. No need to pass it as an argument again, right?

File: src/m2n/tests/PointToPointCommunicationTest.cpp [hunk 3]
```diff
@@ -419,6 +419,276 @@ BOOST_AUTO_TEST_CASE(ConnectionTest, * testing::OnSize(4))
...
+BOOST_AUTO_TEST_CASE(P2PMeshBroadcastTest, * testing::OnSize(4))
+{
+
+  assertion(utils::Parallel::getCommunicatorSize() == 4);
+
+  // com::PtrCommunicationFactory cf(new com::SocketCommunicationFactory);
+  com::PtrCommunicationFactory cf(new com::MPIPortsCommunicationFactory); 
+  utils::MasterSlave::_communication = std::make_shared<com::MPIDirectCommunication>();
+  mesh::PtrMesh mesh(new mesh::Mesh("Mesh", 2, true));
+  int dimensions = 2;
```
Comment (uekerman) [inline src/m2n/tests/PointToPointCommunicationTest.cpp]:
```
  int dimensions = 2;
  mesh::PtrMesh mesh(new mesh::Mesh("Mesh", dimensions, true));
```

File: src/m2n/tests/PointToPointCommunicationTest.cpp [hunk 4]
```diff
@@ -419,6 +419,276 @@ BOOST_AUTO_TEST_CASE(ConnectionTest, * testing::OnSize(4))
...
+
+    Eigen::VectorXd position(dimensions);
+    position <<5.5, 0.0;
+    mesh::Vertex& v1 = mesh->createVertex(position);
+    position << 1.0, 2.0;
+    mesh::Vertex& v2 = mesh->createVertex(position);
+    mesh->createEdge(v1, v2);
+
+    mesh->getConnectedRanks().push_back(0);
+//    mesh->getCommunicationMap()[1].push_back(-1);
```
Comment (uekerman) [inline src/m2n/tests/PointToPointCommunicationTest.cpp]: Please remove

File: src/m2n/tests/PointToPointCommunicationTest.cpp [hunk 5]
```diff
@@ -419,6 +419,276 @@ BOOST_AUTO_TEST_CASE(ConnectionTest, * testing::OnSize(4))
...
+
+    break;
+  }
+  }
+
+  m2n::PointToPointCommunication c(cf, mesh);
+
+  if (utils::Parallel::getProcessRank() < 2) {
+  
+    c.requestPreConnection("Solid", "Fluid");
```
Comment (uekerman) [inline src/m2n/tests/PointToPointCommunicationTest.cpp]: A better unit test design would probably be to not use this functionality here, but to create the pre-connections manually.

File: src/m2n/tests/PointToPointCommunicationTest.cpp [hunk 6]
```diff
@@ -419,6 +419,276 @@ BOOST_AUTO_TEST_CASE(ConnectionTest, * testing::OnSize(4))
...
+    c.acceptPreConnection("Solid", "Fluid");    
+    c.broadcastReceiveMesh(*mesh);
+
+      if(utils::Parallel::getProcessRank() ==2 )
+      {     
+        BOOST_TEST(mesh->vertices().size()==4);
+        BOOST_TEST(mesh->vertices()[2].getCoords()[0]==5.50);
+        BOOST_TEST(mesh->vertices()[2].getCoords()[1]==0.0);
+        BOOST_TEST(mesh->vertices()[3].getCoords()[0]==1.0);
+        BOOST_TEST(mesh->vertices()[3].getCoords()[1]==2.0);
```
Comment (uekerman) [inline src/m2n/tests/PointToPointCommunicationTest.cpp]: What happens to vertices 0 and 1?
Can you please document the expected outcome of the test?

File: src/m2n/tests/PointToPointCommunicationTest.cpp [hunk 7]
```diff
@@ -419,6 +419,276 @@ BOOST_AUTO_TEST_CASE(ConnectionTest, * testing::OnSize(4))
   
 }
 
+BOOST_AUTO_TEST_CASE(P2PMeshBroadcastTest, * testing::OnSize(4))
+{
+
+  assertion(utils::Parallel::getCommunicatorSize() == 4);
+
+  // com::PtrCommunicationFactory cf(new com::SocketCommunicationFactory);
+  com::PtrCommunicationFactory cf(new com::MPIPortsCommunicationFactory);
```
Comment (uekerman) [inline src/m2n/tests/PointToPointCommunicationTest.cpp]: If you use MPI Ports you need to define the MPI Ports label `* boost::unit_test::label("MPI_Ports")`. As we exclude those tests for many systems though, it's probably better to use the sockets communication.

File: src/m2n/tests/PointToPointCommunicationTest.cpp [hunk 8]
```diff
@@ -419,6 +419,276 @@ BOOST_AUTO_TEST_CASE(ConnectionTest, * testing::OnSize(4))
...
+
+    utils::MasterSlave::_communication->requestConnection("Fluid.Master", "Fluid.Slave", 0, 1);
+    
+    mesh->getConnectedRanks().push_back(1);    
+    localCommunicationMap[0].push_back(112);
+    localCommunicationMap[0].push_back(1122);
+    localCommunicationMap[0].push_back(11222);
+    localCommunicationMap[1].push_back(113);
+    localCommunicationMap[1].push_back(1133);
+    localCommunicationMap[1].push_back(11333);
```
Comment (uekerman) [inline src/m2n/tests/PointToPointCommunicationTest.cpp]: Is `11333` some sort of decimal encoding or do you really want to use a vertex ID of 11333?

File: src/partition/tests/ProvidedBoundingBoxTest.cpp [hunk 9]
```diff
@@ -357,6 +417,130 @@ BOOST_AUTO_TEST_CASE(TestComputeBoundingBox, * testing::OnSize(4))
...
+  }
+  }
+
+  mesh->computeState();
+
+  // create communicatror for master com and bb exchange/com/initial com_map
+  com::PtrCommunication participantCom1 = com::PtrCommunication(new com::SocketCommunication());
+  m2n::DistributedComFactory::SharedPointer distrFactory = m2n::DistributedComFactory::SharedPointer(new m2n::GatherScatterComFactory(participantCom1));
+  m2n::PtrM2N m2n = m2n::PtrM2N(new m2n::M2N(participantCom1, distrFactory));
+  setupM2NEnvironment(m2n);
```
Comment (uekerman) [inline src/partition/tests/ProvidedBoundingBoxTest.cpp]: Why do you need all that?

File: src/partition/tests/ProvidedBoundingBoxTest.cpp [hunk 10]
```diff
@@ -357,6 +417,130 @@ BOOST_AUTO_TEST_CASE(TestComputeBoundingBox, * testing::OnSize(4))
   tearDownParallelEnvironment();
 }
 
+BOOST_AUTO_TEST_CASE(TestProvidedReceivedPartitionCommunicate, * testing::OnSize(4))
```
Comment (uekerman) [inline src/partition/tests/ProvidedBoundingBoxTest.cpp]: What about `TestCommunicateLocalMeshPartitions`?

File: src/m2n/GatherScatterCommunication.cpp [hunk 11]
```diff
@@ -178,5 +178,17 @@ void GatherScatterCommunication::broadcastSend(double &itemToSend)
 void GatherScatterCommunication::broadcastReceive(double &itemToReceive)
 {}
 
+void GatherScatterCommunication::broadcastSendMesh(mesh::Mesh &mesh)
+{}
+
```
[the remark this answers is not in the document]
Comment (atotoun) [reply src/m2n/GatherScatterCommunication.cpp]: done!

File: src/m2n/M2N.cpp [hunk 12]
```diff
@@ -103,6 +103,32 @@ void M2N::requestSlavesConnection(
   assertion(_areSlavesConnected);
 }
 
+void M2N::acceptSlavesPreConnection(
+    const std::string &acceptorName,
+    const std::string &requesterName)
+{
+  TRACE(acceptorName, requesterName);
+  _areSlavesConnected = true;
```
[the remark this answers is not in the document]
Comment (atotoun) [reply src/m2n/M2N.cpp]: in the previous approach, we have this at both places, i thought maybe there is a reason for this! Should I remove it from accept/requestPreConnection in the pointtopointcommunication.cpp class?

File: src/m2n/tests/PointToPointCommunicationTest.cpp (hunk 3 above)
Comment (atotoun) [reply src/m2n/tests/PointToPointCommunicationTest.cpp]: done!

File: src/m2n/M2N.cpp (hunk 1 above)
Comment (atotoun) [reply src/m2n/M2N.cpp]: done!

File: src/m2n/PointToPointCommunication.cpp (hunk 2 above)
Comment (atotoun) [reply src/m2n/PointToPointCommunication.cpp]: I am not sure if it's the same one. we give this mesh from partion mackage to m2n.cpp and from there we give it to methods in  the PointToPointCommunication.cpp class.

File: src/m2n/M2N.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (uekerman) [reply src/m2n/M2N.cpp]: Well, we have to decide whether this means that the channels between slaves are established or whether the complete communication is set up (including the vertex information). It should be consistent for both layers, `M2N` and `PointToPointCommunication`.

File: src/m2n/M2N.cpp (hunk 1 above)
Comment (uekerman) [reply src/m2n/M2N.cpp]: The error message makes no sense here, probably a copy-and-paste mistake.

File: src/m2n/PointToPointCommunication.cpp (hunk 2 above)
Comment (uekerman) [reply src/m2n/PointToPointCommunication.cpp]: It should be the same mesh, no need to duplicate that.

File: src/m2n/tests/PointToPointCommunicationTest.cpp (hunk 8 above)
Comment (uekerman) [reply src/m2n/tests/PointToPointCommunicationTest.cpp]: Yes, this is what I meant by "encoding". Can you please document briefly?

File: src/partition/tests/ProvidedBoundingBoxTest.cpp (hunk 9 above)
Comment (uekerman) [reply src/partition/tests/ProvidedBoundingBoxTest.cpp]: `setupM2NEnvironment(m2n)` does not make use of the `m2n`. Also, the name is misleading as only the master-slave communication on each side is set up, not the the m2n communication. I think you can remove all that.

Comment (atotoun) [reply src/partition/tests/ProvidedBoundingBoxTest.cpp]: right! removed!

File: src/m2n/tests/PointToPointCommunicationTest.cpp (hunk 8 above)
Comment (atotoun) [reply src/m2n/tests/PointToPointCommunicationTest.cpp]: done!

File: src/versions.hpp [hunk 13]
```diff
@@ -0,0 +1,8 @@
+#pragma once
```
Comment (fsimonis) [inline src/versions.hpp]: Please remove this file.

File: src/m2n/DistributedCommunication.hpp [hunk 14]
```diff
@@ -117,6 +119,28 @@ class DistributedCommunication
...
+  /**
+   * @brief All ranks receive mesh partition from remote local ranks.
+   */
+  virtual void broadcastReceiveMesh() = 0;
+
+  /**
+   *  All ranks Send their local communication maps to connected ranks
+   */
+  virtual void broadcastSendLCM(
+    std::map<int, std::vector<int>> &localCommunicationMap)=0;
```
Comment (fsimonis) [inline src/m2n/DistributedCommunication.hpp]: Please create a type alias to improve readability. Applies also for below. 

```cpp
using LocalCommunicationMap = std::map<int, std::vector<int>>;
```

**Comment by atotoun** [comment]
@fsimonis changes done, please have a look.

File: src/m2n/DistributedCommunication.hpp [hunk 15]
```diff
@@ -117,6 +120,33 @@ class DistributedCommunication
...
+  
+  /**
+   * @brief All ranks receive mesh partition from remote local ranks.
+   */
+  virtual void broadcastReceiveMesh() = 0;
+
+  /*
+   * A mapping from remote local ranks to the IDs that must be communicated
+   */
+  using communicationMap = std::map<int, std::vector<int>>;
```
Comment (fsimonis) [inline src/m2n/DistributedCommunication.hpp]: Types (names and aliases) are capitalised.

```suggestion
  using CommunicationMap = std::map<int, std::vector<int>>;
```

Comment (atotoun) [reply src/m2n/DistributedCommunication.hpp]: @fsimonis Yes, right. changed!

**Pull request merged**